### ISSUE TITLE
Effects should be scheduled as a macrotask to adapt behavior to React

### ISSUE THREAD

**Issue by clebert** [clebert/batis#11]
```js
function Test() {
  const [state, setState] = React.useState("a");

  if (state === "a") {
    setState("b");

    setTimeout(() => {
      console.log("macrotask1");
    }, 0);

    Promise.resolve()
      .then(() => console.log("microtask1"))
      .then(() => console.log("microtask2"));

    requestAnimationFrame(() => console.log("animationFrame1"));
  }

  React.useEffect(() => {
    console.log("effect1", state);
  }, []);

  React.useEffect(() => {
    console.log("effect2", state);
  });

  React.useLayoutEffect(() => {
    console.log("layoutEffect1", state);
  }, []);

  React.useLayoutEffect(() => {
    console.log("layoutEffect2", state);
  });

  if (state === "a") {
    setTimeout(() => {
      console.log("macrotask2");
    }, 0);

    requestAnimationFrame(() => console.log("animationFrame2"));
  }

  console.log("render", state);

  return <div>{state}</div>;
}
```

Console logs:

```
"render", "a"
"render", "b"
"layoutEffect1", "b"
"layoutEffect2", "b"
"microtask1"
"microtask2"
"animationFrame1"
"animationFrame2"
"macrotask1"
"macrotask2"
"effect1", "b"
"effect2", "b"
```

**Comment by clebert** [comment]
Strange observation in the behavior of the React implementation:

```js
function Test() {
  const [state, setState] = React.useState("a");

  if (state === "a") {
    setState("b");

    setTimeout(() => {
      console.log("macrotask1");
    }, 0);

    Promise.resolve()
      .then(() => console.log("microtask1"))
      .then(() => console.log("microtask2"));

    requestAnimationFrame(() => console.log("animationFrame1"));
  }

  React.useEffect(() => {
    setState('d');

    console.log("effect1", state);
  }, []);

  React.useEffect(() => {
    console.log("effect2", state);
  });

  React.useLayoutEffect(() => {
    setState('c');

    console.log("layoutEffect1", state);
  }, []);

  React.useLayoutEffect(() => {
    console.log("layoutEffect2", state);
  });

  if (state === "a") {
    setTimeout(() => {
      console.log("macrotask2");
    }, 0);

    requestAnimationFrame(() => console.log("animationFrame2"));
  }

  console.log("render", state);

  return <div>{state}</div>;
}
```

Console logs:

```
"render", "a"
"render", "b"
"layoutEffect1", "b"
"layoutEffect2", "b"
"effect1", "b"
"effect2", "b"
"render", "d"
"layoutEffect2", "d"
"microtask1"
"microtask2"
"animationFrame1"
"animationFrame2"
"macrotask1"
"macrotask2"
"effect2", "d"
```

**Comment by clebert** [comment]
```js
function Test() {
  const [state, setState] = React.useState("a");

  if (state === "a") {
    setState("b");

    setTimeout(() => {
      console.log("macrotask1");
    }, 0);

    Promise.resolve()
      .then(() => {
        setState('c');
        console.log("microtask1");
      })
      .then(() => console.log("microtask2"));

    requestAnimationFrame(() => console.log("animationFrame1"));
  }

  React.useEffect(() => {
    console.log("effect1", state);
  }, []);

  React.useEffect(() => {
    console.log("effect2", state);
  });

  React.useLayoutEffect(() => {
    console.log("layoutEffect1", state);
  }, []);

  React.useLayoutEffect(() => {
    console.log("layoutEffect2", state);
  });

  if (state === "a") {
    setTimeout(() => {
      console.log("macrotask2");
    }, 0);

    requestAnimationFrame(() => console.log("animationFrame2"));
  }

  console.log("render", state);

  return <div>{state}</div>;
}
```

Console logs:

```
"render", "a"
"render", "b"
"layoutEffect1", "b"
"layoutEffect2", "b"
"effect1", "b"
"effect2", "b"
"render", "c"
"layoutEffect2", "c"
"microtask1"
"microtask2"
"animationFrame1"
"animationFrame2"
"macrotask1"
"macrotask2"
"effect2", "c"
```

**Comment by clebert** [comment]
Since Batis is not intended for UI rendering, I have concluded that it makes no sense to run effects as a macrotask. I have added the following note to the README:

> React executes effects declared using the `useEffect` Hook as a
macrotask so that they do not block browser rendering. The `useLayoutEffect`
Hook acts as the synchronous alternative to allow potential DOM changes before
browser rendering. In Batis, effects are executed always synchronously, so the
special `useLayoutEffect` Hook is not necessary. For compatibility reasons,
`useEffect` could be used as an alias for `useLayoutEffect`.
